src/lib/common/sol-platform-impl-linux-micro.c

```diff
@@ -514,16 +522,137 @@ teardown_pid1(void)
+
+        err = snprintf(arg, sizeof(arg), "%.*s", SOL_STR_SLICE_PRINT(*token));
+        if (err < 0 || err >= (int)sizeof(arg)) {
+            SOL_ERR("Could not format arg");
+            continue;
+        }
+
+        if (streq(arg, SOL_DEBUG_ARG)) {
+            res = true;
+        } else if (strstartswith(arg, SOL_DEBUG_COMM_ARG)) {
+            gdb->comm = strdup(arg);
+            if (res) {
+                SOL_DBG("Got "SOL_DEBUG_ARG " flag and comm: %s", arg);
+                break;
+            }
+        }
+    }
+
+    // if no comm is provided set the default one
+    if (!gdb->comm) {
+        gdb->comm = strdup("/dev/tty0");
+        if (!gdb->comm) {
+            SOL_INF("Could not allocate gdbserver comm desc memory");
+            return false;
+        }
+        SOL_INF("No debug comm provided, using default one (/dev/tty0)");
+    }
+
+    sol_file_reader_close(fr);
+    return res;
+}
+
+static void
+on_fork(void *data)
+{
+    int err;
+    unsigned int i;
+    char pids[PATH_MAX];
+    struct gdbserver *gdb = data;
+
+    const char *paths[] = {
+        "/usr/bin/gdbserver",
+        "/bin/gdbserver",
+    };
+
+    err = snprintf(pids, sizeof(pids), "%d", gdb->pid);
+    if (err < 0 || err >= (int)sizeof(pids)) {
+        SOL_ERR("Could not format pid string to attach gdbserver");
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    if (setenv("SOL_LOAD_INITIAL_SERVICES", "1", 1) == -1) {
+        SOL_ERR("Could not set SOL_LOAD_INITIAL_SERVICES");
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    for (i = 0; i < ARRAY_SIZE(paths); i++) {
+        if (execl(paths[i], paths[i], "--attach", gdb->comm, pids, NULL) == -1) {
+            SOL_ERR("failed to exec gdbserver - %s", sol_util_strerrora(errno));
+            continue;
+        }
+    }
+
+    SOL_ERR("Failed to exec() gdbserver");
+    sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+}
+
+static void
+on_fork_exit(void *data, uint64_t pid, int status)
+{
+    SOL_ERR("gdbserver exited... %" PRIu64, pid);
+}
+
 int
 sol_platform_impl_init(void)
 {
     bool want_load_initial_services = false;
+    struct gdbserver gdb;
+    pid_t pid = getpid();
 
-    if (getpid() == 1 && getppid() == 0) {
+    if (pid == 1 && getppid() == 0) {
         int err = setup_pid1();
         SOL_INT_CHECK(err, < 0, err);
 
         want_load_initial_services = true;
+
+        gdb.pid = pid;
+        gdb.comm = NULL;
+
+        if (sol_platform_linux_micro_should_debug(&gdb)) {
+            struct sol_platform_linux_fork_run *fork;
+
+            fork = sol_platform_linux_fork_run(on_fork, on_fork_exit, &gdb);
+            if (!fork) {
+                SOL_ERR("Could not launch gdbserver - %s",
+                    sol_util_strerrora(errno));
+            } else {
+                while (true) {
+                    int status;
+                    pid_t child = wait(&status);
+
+                    SOL_DBG("child pid=%" PRIu64 " status=%d",
```

Comment:
ah, and one idea would be to turn this `fork() -> exec(gdbserver)` into a loop. Then if gdbserver crashes or finishes we restart it. I guess it's more useful than sit in a `wait()` forever (or constant error if there are no more PIDs to wait).

```c
static void
gdb_wait(pid_t gdb_pid)
{
    while (true) {
        int status;
        pid_t child = wait(&status);
        if (child < 0) {
            if (errno == EINTR)
                continue;
            SOL_WRN("wait() failed: %s", sol_util_strerrnoa(errno));
            return; // no more child process, they are all dead, restart gdbserver -- it is just to be complete, it should never happen as we check if gdbserver is dead to restart it
        } else {
            SOL_DBG("child pid=%" PRIu64 " status=%d", (uint64_t)child, status);
            if (child == gdb_pid) {
                SOL_INF("gdbserver exited, restart it");
                return;
            }
        }
    }
}

static void
gdb_exec(const char *gdb_comm)
{
    const char **argv;
    int argv;
    const char **args;
    const char *paths[] = {
        "/usr/bin/gdbserver",
       "/bin/gdbserver",
    };

    if (setenv("SOL_LOAD_INITIAL_SERVICES", "1", 1) == -1) {
        SOL_ERR("Could not set SOL_LOAD_INITIAL_SERVICES");
        _exit(EXIT_FAILURE);
    }

    argc = sol_argc();
    argv = sol_argv();
    if (argc < 1 || !argv || !argv[0]) {
        SOL_ERR("Invalid argc=%d, argv=%p, argv[0]=%p", argc, argv, argv ? argv[0] : NULL);
        _exit(EXIT_FAILURE);
    }

    args = alloca((argc + 3) * sizeof(char *));
    args[1] = gdb_comm;
    memcpy(args + 2, argv, argc * sizeof(char *));
    args[argc + 2] = NULL;

    for (i = 0; i < ARRAY_SIZE(paths); i++) {
        args[0] = paths[i];
        if (execv(paths[i], args) == -1)
            SOL_DBG("failed to exec %s - %s", paths[i], sol_util_strerrora(errno)); // not an error, we'll fallback...
    }
    
   SOL_WRN("no gdb server found, execute the application without it");
   execv(argv[0], argv);
   SOL_CRIT("could not execute the application %s: %s", argv[0], sol_util_strerrora(errno));
   _exit(EXIT_FAILURE);
}

static void gdb_debug(const char *gdb_comm) SOL_ATTR_NORETURN;

static void
gdb_debug(const char *gdb_comm)
{
    while (true) {
        pid_t gdb_pid = fork();
        if (gdb_pid < 0) {
            SOL_WRN("could not fork(): %s", sol_util_strerrnoa(errno));
            sleep(1); // give the system some time to breath
        } else if (gdb_pid > 0) {
            gdb_wait(gdb_pid);
        } else {
            gdb_exec(gdb_comm);
        }
    }
}

int
sol_platform_impl_init(void)
{
...
char *gdb_comm;
if (should_gdb_debug(&gdb_comm)) // no need to prefix the static func names, not being used in this file
    gdb_debug(&gdb_comm);
...
}
```